Initial implementation of requestIdleCallback on Android

janicduplessis opened the pull request:
This is a follow up of the work by @brentvatne in #5052. This addresses the feedback by @astreet.

- Uses ReactChoreographer with a new callback type
- Callback dispatch logic moved to JS
- Only calls into JS when needed, when there are pending callbacks, it even removes the Choreographer listener when no JS context listen for idle events.

** Test plan **
Tested by running a background task that burns all remaining idle time (see new UIExplorer example) and made sure that UI and JS fps stayed near 60 on a real device (Nexus 6) with dev mode disabled. Also tried adding a JS driven animation and it stayed smooth.

Tested that native only calls into JS when there are pending idle callbacks.

Also tested that timers are executed before idle callback.
```
requestIdleCallback(() => console.log(1));
setTimeout(() => console.log(2), 100);
burnCPU(1000);
// 2
// 1
```

I did *not* test with webworkers but it should work as I'm using executor tokens.

janicduplessis commented:
Few concerns for discussion:
- This assumes that frames are 17ms long, maybe we could get this value from the display refresh rate.
- Has an arbitrary 10ms time left threshold to trigger the callback, maybe this will be too much for slow android devices?

ide commented:
> This assumes that frames are 17ms long, maybe we could get this value from the display refresh rate.

This would be nice to have in the future, where I can see us dropping the frame rate from 60fps to 30fps intentionally when we discover that the device cannot handle the current workload. This is what Chrome does for example.

brentvatne commented on Libraries/JavaScriptAppEngine/System/JSTimers/JSTimersExecution.js:
```diff
@@ -108,6 +117,34 @@ var JSTimersExecution = {
 
+  callIdleCallbacks: function(frameTime) {
+    // Arbitrary threshold: 10ms remaining in frame.
+    if (17 - (performanceNow() - frameTime) < 10) {
```
@sahrens @astreet - the threshold to be used here in the spec says it is determined by the user agent, we could definitely use something more sophisticated than this but for now this seems fine

sahrens replied:
@sebmarkbage: do you have any thoughts on this?

sebmarkbage replied:
The only key here is that this should be *longer* than a frame. It should be possible to go beyond 17ms in an idle callback while still hitting every frame deadline. E.g. 5ms to process the current frame, 20ms to process the idle callback, 5ms to process the next frame. Still hits both deadlines.

If you don't call it often enough, then you're leaving CPU cycles unused. Only calling this if there is 10ms left is way too restrictive because there is too much headroom left out. Remember, even if it goes beyond the frame deadline, it doesn't matter. This should be like 1ms or even less.

sebmarkbage replied:
Note that the new React Fiber reconciler for DOM does __all__ reconciliation except for animations using requestIdleCallback. https://github.com/facebook/react/blob/cf259a4ff82bb6248b41392e821d2822f9ce46c6/src/renderers/shared/fiber/ReactFiberScheduler.js#L199

So it is very important that it gets enough cycles. It is a misnomer to think that it is only for background work.

janicduplessis replied:
Let's call it whenever there is some time left in the frame for now. We can iterate on it and improve the scheduling later. The new React reconciler will definitely be a great test for this and we'll be able to optimize for it.

astreet replied:
Taking time at the beginning of the frame (by overrunning into the next frame) seems pretty dangerous to me. I mean, generally trying to schedule at the granularity of half-a-frame without hooks into the actual scheduler seems dangerous to me, but *assuming* you're going to have idle time is even more so :P

sebmarkbage replied:
Note that scheduling running into the next frame is the _only_ reason this API exists at all in browsers. If you wanted to ensure to have ample time at the end of a frame you can just use `setTimeout`.

If you assume that you take close to entire frame to update a frame, you are already in very dangerous territory because you're likely to run across - because of a miscalculation, slower device or OS scheduling interrupting you. So the argument goes two ways. If your idle time is too risky at the edges then so is the frame calculation.

In practice you don't need that much time to calculate the next frame, outside of compositing. This heuristic seems to assume that it normally takes less than 6ms. If that's the case then there is plenty of headroom even if you run over.

This is the only reason this API is competitive with threads.

Btw, Chrome by default gives you 50ms before any animation starts. Because dropping a frame on the initial interaction is NBD. When there are animations happening it cuts down but still leaving you ample time to utilize as many cycles as possible.

sebmarkbage replied:
If you assume that your normal workload is before this then you certainly will be close to the 16ms in which case eating into a frame would be dangerous. However that's just an artifact of so many events being processed before the animation is flushed. Most of that event handling such as handling network responses should actually be considered idle work.

One thing that made me shift my mental model was to think of these as requestNormalCallback instead of Idle. Because it is really not suppose to be just background work but most work.

brentvatne commented on flow/global.js:
```diff
@@ -0,0 +1,13 @@
+
+// Add missing flow definitions, can be removed when included in flow.
+declare function requestIdleCallback(callback: any): number;
+declare function cancelIdleCallback(idleCallbackId?: any): void;
```
I think the idleCallbackId should be type `number` and not optional -- this may have been my mistake, oops.

janicduplessis replied:
I adapted it from the setTimeout definition from flow, I think passing anything to cancelIdleCallback is fine it will just be a no-op.

sahrens replied:
Yeah, we definitely want it optional. It's a pain to check before calling cancel.

brentvatne replied:
ah, gotcha :+1:

satya164 replied:
Can we make it `idleCallbackId: ?number` instead of `idleCallbackId?: any`?

`idleCallbackId?: any` doesn't make much sense since it implies that you can just call `cancelIdleCallback` without any argument. `idleCallbackId: ?number` is better since you have to pass a number as argument, but you can pass null as well, so no need to check.

brentvatne commented:
This looks good to me, but I don't know nearly as much about Android or the react-native Android internals as @astreet so let's wait for his feedback :)

sebmarkbage commented on Libraries/JavaScriptAppEngine/System/JSTimers/JSTimersExecution.js:
```diff
@@ -70,6 +73,12 @@ var JSTimersExecution = {
+      } else if (type === JSTimersExecution.Type.requestIdleCallback) {
+        callback({
+          timeRemaining: function() {
+            return Math.max(0, 17 - (performanceNow() - frameTime));
```
Ideally this would be longer if the system is idle and then read out to shared mutable memory. It can then return zero when there are pending touches or new input to process. That way code can synchronously for as long as there is no new input to process.

janicduplessis replied:
@astreet Do we have any way to do this with our current JSC setup? We'll want to know if there are any pending calls from native synchronously.

ide replied:
Just want to chime in and suggest adding a TODO for now if synchronous JS execution isn't yet figured out. For example, if we rely on synchronous JS, the Chrome debugger won't work and therefore you can't debug rIC code. (This limitation seems OK for more obscure libraries, but for a core API like rIC I think we want it to be debuggable.)

janicduplessis replied:
Yeah I dont think we should implement this in this PR either, as for remote debugging we can probably just skip this check when debugging in chrome and sync calls are not possible as it is more of an optimisation than something that is required.

sebmarkbage replied:
I don't think the debugging argument is significant here because this whole API is a massive race condition regardless. There's no reliable break debugging strategy here anyway. A TODO would be good though.

astreet replied:
Yeah, the way we'd want to do this at first is to check if there are any pending JS calls to process. We can/do keep track of this in the native bridge, but this isn't exposed right now.

As far as debugging, the issue with the sync hook is less debugging this in particular and more being able to debug an application that uses this API at all: an app that uses this API cannot run in Chrome.

Internally, we have a lot of reasons for wanting an on-device debugger but are still looking for someone to take the project on.

sebmarkbage replied:
We don't need to wait for an on device debugger though. If the hook is only exposed on device then we just use feature detection on the JS side. It is not strictly necessary to work, it is only an optimization.

astreet replied:
Ah yeah, if we have a separate implementation when not on device we should be fine, though obviously that's not ideal for a number of reasons (but as you say, not a blocker)

astreet commented on ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactChoreographer.java:
```diff
@@ -46,6 +46,11 @@
     TIMERS_EVENTS(3),
+
+    /**
+     * Event used to dispatch the idle callback. Called after all other work is done for the frame.
```
I would rephrase this to say that these are called when there is no other UI work that will be dispatched to JS this frame: after this is called, there is still the layout/draw pass of the frame on the UI thread.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -275,4 +377,28 @@ public void deleteTimer(ExecutorToken executorToken, int timerId) {
   }
+
+  @ReactMethod
+  public void sendIdleEvents(ExecutorToken executorToken, boolean shouldSend) {
```
setShouldSendIdleEvents

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +112,78 @@ public void doFrame(long frameTimeNanos) {
+
+      List<ExecutorToken> contextsToCall;
+      synchronized (mIdleCallbackGuard) {
+        contextsToCall = new ArrayList<>(mSendIdleEventsExecutorTokens);
```
let's try to avoid this allocation on every frame. We can just keep one array and copy into it

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +112,78 @@ public void doFrame(long frameTimeNanos) {
+      if (mCanceled) {
+        return;
+      }
+
```
we can pretty easily exit early here if we're already past our deadline for the frame (we probably want to share that 10ms constant between native and JS so that it's consistent)

astreet commented:
Native implementation looks good, see inlines :)

ide commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +112,78 @@ public void doFrame(long frameTimeNanos) {
+  }
+
+  private class IdleCallbackRunnable implements Runnable {
+    private volatile boolean mCanceled = false;
```
nit: cancelled seems to be more popular than canceled in the RN code. both are correct thought =P

janicduplessis commented:
Addressed feedback from @astreet and @ide. Also added some TODOs for future PRs where optimizations would be possible as mentioned by @sebmarkbage. I guess looking at what browsers do will help a lot in improving the scheduling.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +120,84 @@ public void doFrame(long frameTimeNanos) {
+        return;
+      }
+
+      mIdleCallbackContextsToCall.clear();
```
It looks like this isn't necessary anymore

astreet replied:
Err, whoops, I see what you're doing now

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +120,84 @@ public void doFrame(long frameTimeNanos) {
+  private boolean mFrameIdleCallbackPosted = false;
+  private final Set<ExecutorToken> mSendIdleEventsExecutorTokens;
+  // Temporary array used to dipatch idle callbacks on the JS thread.
+  private final List<ExecutorToken> mIdleCallbackContextsToCall;
```
Since this is only used by the callback can you move it as an instance var there?

janicduplessis replied:
This is not possible since we create a new runnable every frame. I don't know if there is a way to avoid creating this runnable every frame while still sending only the latest event to JS.

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -28,17 +20,38 @@
   OnExecutorUnregisteredListener {
 
+  // The minimum time in milliseconds left in the frame to call idle callbacks.
+  private static final float IDLE_CALLBACK_FRAME_DEADLINE = 1.f;
```
Add _MS to both of these

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -28,17 +20,38 @@
+  // The total duration of a frame in milliseconds, this assumes that devices run at 60 fps.
+  // TODO: Lower frame duration on devices that are too slow to run consistently
+  // at 60 fps.
+  private static final float FRAME_DURATION = 17.f;
```
1000 / 60.f then?

janicduplessis replied:
Sure, I left it as 17 since we round the other MS values (by using longs). Should we try to keep decimals for the frame time then?

astreet commented on ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java:
```diff
@@ -107,13 +120,84 @@ public void doFrame(long frameTimeNanos) {
+      }
+
+      mCurrentIdleCallbackRunnable = new IdleCallbackRunnable(frameTimeNanos);
+      getReactApplicationContext().runOnJSQueueThread(mCurrentIdleCallbackRunnable);
```
Sorry I didn't catch this before: can you remind me why we're posting a runnable to the JS thread that then posts more runnables (via getJSModule(...).callIdleCallbacks)? Why not just callIdleCallbacks here?

janicduplessis replied:
This is to avoid queueing multiple calls to JS if the JS thread is blocked for multiple frames. What I'm doing here is every frame I'm sending a new runnable and cancelling the previous one so only the latest will be sent to JS.

astreet replied:
Ok, seems like it shouldn't hurt, especially if we really want this to run when we're idle.

I'm not sure how expensive it is to allocate one of these runnables on each frame, but we could pretty easily recycle them.

astreet commented:
@facebook-github-bot shipit

janicduplessis commented:
Oups, forgot to fix and run tests after the last changes. Let's try shipping again.

@facebook-github-bot shipit

facebook-github-bot-1 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/526640120858545/int_phab) to review.
